contracts/HoneyFarm.sol

```diff
@@ -218,54 +237,53 @@ contract HoneyFarm is Ownable, ERC721 {
         );
+        require(_pools.contains(address(poolToken)), "HF: Non-existant pool");
         PoolInfo storage pool = poolInfo[poolToken];
         updatePool(poolToken);
-        pool.lpToken.safeTransferFrom(
+        poolToken.safeTransferFrom(
             address(msg.sender),
             address(this),
             amount
         );
         uint256 newDepositId = totalDeposits++;
-        _safeMint(msg.sender, newDepositId);
-        DepositInfo memory newDeposit = DepositInfo({
+        uint256 newShares = amount.mul(getTimeMultiple(unlockTime));
+        pool.totalShares = pool.totalShares.add(newShares);
+        depositInfo[newDepositId] = DepositInfo({
             amount: amount,
-            rewardDebt: amount.mul().div(SCALE),
+            rewardDebt: amount.mul(pool.accHsfPerShare).div(SCALE),
             unlockTime: unlockTime,
-
-
+            rewardShare: newShares,
+            pool: poolToken
         });
-        emit Deposit(msg.sender, _pid, _amount);
+        _safeMint(msg.sender, newDepositId);
     }
 
     // Withdraw LP tokens from MasterChef.
-    function withdraw(uint256 _pid, uint256 _amount) public {
-        PoolInfo storage pool = poolInfo[_pid];
-        UserInfo storage user = userInfo[_pid][msg.sender];
-        require(user.amount >= _amount, "withdraw: not good");
-        updatePool(_pid);
+    function closeDeposit(uint256 depositId) public {
+        require(ownerOf(depositId) == msg.sender, "HF: Must be owner to withdraw");
+        DepositInfo storage deposit = depositInfo[depositId];
+        require(
+            deposit.unlockTime == 0 ||
+            deposit.unlockTime <= block.timestamp ||
+            contractDisabled,
+            "HF: Deposit still locked"
+        );
+        IERC20 poolToken = deposit.pool;
+        PoolInfo storage pool = poolInfo[poolToken];
+        updatePool(poolToken);
+
         uint256 pending =
-            user.amount.mul(pool.accSushiPerShare).div(1e12).sub(
-                user.rewardDebt
+            deposit.rewardShare.mul(pool.accHsfPerShare).div(SCALE).sub(
+                deposit.rewardDebt
             );
-        safeSushiTransfer(msg.sender, pending);
-        user.amount = user.amount.sub(_amount);
-        user.rewardDebt = user.amount.mul(pool.accSushiPerShare).div(1e12);
-        pool.lpToken.safeTransfer(address(msg.sender), _amount);
-        emit Withdraw(msg.sender, _pid, _amount);
-    }
 
-    // Withdraw without caring about rewards. EMERGENCY ONLY.
-    function emergencyWithdraw(uint256 _pid) public {
-        PoolInfo storage pool = poolInfo[_pid];
-        UserInfo storage user = userInfo[_pid][msg.sender];
-        pool.lpToken.safeTransfer(address(msg.sender), user.amount);
-        emit EmergencyWithdraw(msg.sender, _pid, user.amount);
-        user.amount = 0;
-        user.rewardDebt = 0;
+        _safeHsfTransfer(msg.sender, pending);
+        poolToken.safeTransfer(msg.sender, deposit.amount);
+        _burn(depositId);
```

Comment:
maybe it would safer to burn the depositId before the transfers

Reply:
Good idea, will rearrange.